src/extra/gpkg.jl

```diff
@@ -0,0 +1,1054 @@
+    end
+  end
+  
+  header = GeoPackageBinaryHeader(
+    magic, version, flags, srs_id,
+    min_x, max_x, min_y, max_y,
+    min_z, max_z, min_m, max_m
+  )
+  
+  return header, position(io)
+end
+
+function parse_wkb_point(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool, crs_type)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  
+  x = read_value(read(io, Float64))
+  y = read_value(read(io, Float64))
+  
+  if isnan(x) && isnan(y)
+    return nothing
+  end
+  
+  coords_list = [x, y]
+  
+  if has_z
+    z = read_value(read(io, Float64))
+    push!(coords_list, z)
+  end
+  
+  if has_m
+    m = read_value(read(io, Float64))
+  end
+  
+  # Create point with appropriate CRS
+  if crs_type <: LatLon
+    if length(coords_list) == 2
+      return Point(crs_type(y, x))  # LatLon(lat, lon) - note swapped order
+    else
+      return Point(crs_type(y, x, coords_list[3]))
+    end
+  else
+    return Point(coords_list...)
+  end
+end
+
+function parse_wkb_linestring(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool, crs_type)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  
+  num_points = read_value(read(io, UInt32))
+  
+  # Create the first point to determine the concrete type
+  x = read_value(read(io, Float64))
+  y = read_value(read(io, Float64))
+  
+  coords_list = [x, y]
+  
+  if has_z
+    z = read_value(read(io, Float64))
+    push!(coords_list, z)
+  end
+  
+  if has_m
+    m = read_value(read(io, Float64))
+  end
+  
+  # Create first point with appropriate CRS
+  first_point = if crs_type <: LatLon
+    if length(coords_list) == 2
+      Point(crs_type(y, x))  # LatLon(lat, lon) - note swapped order
+    else
+      Point(crs_type(y, x, coords_list[3]))
+    end
+  else
+    Point(coords_list...)
+  end
+  
+  # Initialize typed array with the correct concrete type
+  points = [first_point]
+  
+  # Parse remaining points
+  for i in 2:num_points
+    x = read_value(read(io, Float64))
+    y = read_value(read(io, Float64))
+    
+    coords_list = [x, y]
+    
+    if has_z
+      z = read_value(read(io, Float64))
+      push!(coords_list, z)
+    end
+    
+    if has_m
+      m = read_value(read(io, Float64))
+    end
+    
+    # Create point with appropriate CRS
+    if crs_type <: LatLon
+      if length(coords_list) == 2
+        push!(points, Point(crs_type(y, x)))  # LatLon(lat, lon) - note swapped order
+      else
+        push!(points, Point(crs_type(y, x, coords_list[3])))
+      end
+    else
+      push!(points, Point(coords_list...))
+    end
+  end
+  
+  # Check if the linestring is closed (first point equals last point)
+  if length(points) >= 2 && points[1] == points[end]
+    # It's a closed linestring, return a Ring (exclude the duplicate last point)
+    return Ring(points[1:end-1])
+  else
+    # It's an open linestring, return a Rope
+    return Rope(points)
+  end
+end
+
+function parse_wkb_linestring_as_rope(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool, crs_type)
```

Comment:
We don't need two copies of the function to load Rope vs. Ring. They are basically the same, except that the latter is semantically closed. If you create a Rope or a Ring from the same list of vertices, they will have the expected behavior.